src/module.api/poolswap.pathfinding.service.ts

```diff
@@ -0,0 +1,233 @@
+import { Injectable, NotFoundException } from '@nestjs/common'
+import { UndirectedGraph } from 'graphology'
+import {
+  AllSwappableTokensResult,
+  BestSwapPathResult,
+  SwapPathPoolPair, SwapPathsResult,
+  TokenIdentifier
+} from '@whale-api-client/api/poolpairs'
+import { PoolPairToken, PoolPairTokenMapper } from '@src/module.model/pool.pair.token'
+import { DeFiDCache } from '@src/module.api/cache/defid.cache'
+import { Interval } from '@nestjs/schedule'
+import BigNumber from 'bignumber.js'
+import { allSimplePaths } from 'graphology-simple-path'
+import { parseDisplaySymbol } from '@src/module.api/token.controller'
+import { PoolPairInfo } from '@defichain/jellyfish-api-core/dist/category/poolpair'
+import { connectedComponents } from 'graphology-components'
+
+@Injectable()
+export class PoolSwapPathFindingService {
+  tokenGraph: UndirectedGraph = new UndirectedGraph()
+  tokensToSwappableTokens = new Map<TokenIdentifier['id'], TokenIdentifier[]>()
+
+  constructor (
+    protected readonly poolPairTokenMapper: PoolPairTokenMapper,
+    protected readonly deFiDCache: DeFiDCache
+  ) {
+  }
+
+  @Interval(120_000) // 120s
+  async syncTokenGraph (): Promise<void> {
+    const poolPairTokens = await this.poolPairTokenMapper.list(200)
+    await this.addTokensAndConnectionsToGraph(poolPairTokens)
+    await this.updateTokensToSwappableTokens()
+  }
+
+  async getAllSwappableTokens (tokenId: string): Promise<AllSwappableTokensResult> {
+    await this.syncTokenGraphIfEmpty()
+
+    return {
+      fromToken: await this.getTokenIdentifier(tokenId),
+      swappableTokens: this.tokensToSwappableTokens.get(tokenId) ?? []
+    }
+  }
+
+  async getBestPath (fromTokenId: string, toTokenId: string): Promise<BestSwapPathResult> {
+    const { fromToken, toToken, paths } = await this.getAllSwapPaths(fromTokenId, toTokenId)
+
+    let bestPath: SwapPathPoolPair[] = []
+    let bestReturn = new BigNumber(-1)
+
+    for (const path of paths) {
+      const totalReturn = computeReturnInDestinationToken(path, fromTokenId)
+      if (totalReturn.isGreaterThan(bestReturn)) {
+        bestReturn = totalReturn
+        bestPath = path
+      }
+    }
+    return {
+      fromToken: fromToken,
+      toToken: toToken,
+      bestPath: bestPath,
+      estimatedReturn: bestReturn.eq(-1)
+        ? '0'
+        : bestReturn.toFixed(8) // denoted in toToken
```

Comment:
```suggestion
        : bestReturn.dp(8).toFixed() // denoted in toToken
```
I think will be cleaner and more consistent.

Reply:
Interesting - I see .toFixed(8) being used more across whale and jellyfish though

Reply:
Or make the previous line `? '0.00000000'` works too. I guess it is depend on the result consumer, for API, less data can be considered better.

Reply:
Oh my bad, i didn't realise `.dp(8).toFixed()` truncates the leading zeroes

I fixed it to make it all consistently return '0' another way though